### DISCUSSION TITLE
routing to self URI then taking data with useFetch() func

### DISCUSSION THREAD

**Discussion by naoyoshi** [nuxt/nuxt#18084] [Questions]
Hello.
I'm really beginner of Vue.js and Nuxt.js and I've been studying about Vue.js 3 and Nuxt.js 3 from last month.

By the way, here are things which I want to do are below:
1) As a first view, showing all data inside table element, and option data inside select element.
2) To change drop down item.
3) After running change event, routing to self URI with parameter.
4) After receiving parameter, take data to server with useFetch() func and show data inside a table element.

I can check to run about upper 1, 2, and 3.
But I'm at a loss how I code about upper 4.

The below is my sample code:
```
sample.vue

<template>
<div>
<Html>
<Head>
<Title>Test Page</Title>
</Head>
</Html>
<h1>Test Page</h1>

<form role="form">
<select @change="onChange($event)">
<option v-for="option in data.option_lists" :key="option.value" :value="option.value">{{ option.label }}</option>
</select>
</form>

<table>
<thead>
<tr>
<th>Date</th>
<th>Category</th>
<th>PDF</th>
<th>Created date</th>
</tr>
</thead>
<tbody>
<tr v-for="item in data.report_lists" :key="item.id">
<th>{{ item.month }}/{{ item.year }}</th>
<th>{{ item.category_name }}</th>
<td><a :href="item.pdf">show</a></td>
</tr>
</tbody>
</table>
</div>
</template>

<script setup>
let category_id = null;

//;;;Here I want to take query parameter like below;;;
//category_id = this.$route.params.category_id;

let param = {
  'category_id': category_id,
};

const { data } = await useFetch('http://localhost:8000/api/foo', {
    params: param
});
</script>
<script>
export default defineComponent({
  methods: {
    onChange: function(event) {
      this.$router.push({
        path: '/sample',
        query: {category_id: event.target.value}
      });
    },
  },
})
</script> 
```

I understood that <script setup> can't run only one time. 
So how do I modify code to achieve what I want to do?

**Comment by viii47** [comment]
Generally speaking you need to return your data

```js
<script setup>
let category_id = null;

//;;;Here I want to take query parameter like below;;;
//category_id = this.$route.params.category_id;

let param = {
  'category_id': category_id,
};

const { data } = await useFetch('http://localhost:8000/api/foo', {
    params: param
});

// here
return { data }
</script>
```


A better approach (imo) is the use of `useState()` and a composable to give you a better accessability to the data.

**Reply by naoyoshi** [reply]
@viii47 
Thank you for reply.

You told me below code.
---
// here
return { data }
---
I tried to code this line.
But below error shows.
[plugin:vite:vue] [@vue/compiler-sfc] 'return' outside of function.

https://v3.nuxtjs.org/docs/usage/data-fetching
In this doc page, there is no sample to explain "return { data }".

Any ideas?

**Comment by naoyoshi** [comment]
I updated my sample code like below.

```
<script>
export default defineComponent({
	methods: {
		onChange(event) {
			fetch('http://localhost:8000/api/foo?category_id='+event.target.value)
			.then(function(res) {
				return res.json()
			})
			.then(function(json) {
				//??? = json.report_lists  # I guess this...  but I don't know what the correct code is?
			});

			this.$router.push({
				path: '/sample',
				query: {category_id: event.target.value}
			});
		},
	},

	beforeRouteUpdate(to, from, next) {
		next();
	},
})
</script>
```

In this code, I can see about request and response data with Fetch API.
Now what I don't know is that how it reactive and refresh "data.report_lists".
